src/java.base/share/classes/sun/security/util/Cache.java

```diff
@@ -246,35 +262,42 @@ public void accept(CacheVisitor<K,V> visitor) {
 
 }
 
-class MemoryCache<K,V> extends Cache<K,V> {
+/**
+ * There are three types of cache.  SoftCacheEntry, HardCacheEntry, and
+ * QueueCacheEntry.  This enum is used to identify the cache's configuration.
+ */
+enum CacheType { SOFT, HARD, CACHE }
 
-    private static final float LOAD_FACTOR = 0.75f;
+class MemoryCache<K,V> extends Cache<K,V> {
 
-    // XXXX
+    // Debugging
     private static final boolean DEBUG = false;
 
     private final Map<K, CacheEntry<K,V>> cacheMap;
     private int maxSize;
     private long lifetime;
     private long nextExpirationTime = Long.MAX_VALUE;
+    private final CacheType cacheType;
 
     // ReferenceQueue is of type V instead of Cache<K,V>
     // to allow SoftCacheEntry to extend SoftReference<V>
     private final ReferenceQueue<V> queue;
 
-    public MemoryCache(boolean soft, int maxSize) {
-        this(soft, maxSize, 0);
+    public MemoryCache(CacheType type, int maxSize) {
+        this(type, maxSize, 0);
     }
 
-    public MemoryCache(boolean soft, int maxSize, int lifetime) {
+    public MemoryCache(CacheType type, int maxSize, int lifetime) {
         this.maxSize = maxSize;
         this.lifetime = lifetime * 1000L;
-        if (soft)
-            this.queue = new ReferenceQueue<>();
-        else
+        cacheType = type;
+        if (cacheType == CacheType.HARD) {
             this.queue = null;
+        } else {
+            this.queue = new ReferenceQueue<>();
+        }
 
-        cacheMap = new LinkedHashMap<>(1, LOAD_FACTOR, true);
+        cacheMap = new ConcurrentHashMap<>();
```

Comment:
With LinkedHashMap we were removing least recently used entries on overflow (see the implementation of `put`); with ConcurrentHashMap we will remove entries in random order. Is that intentional? If it is, you might want to review the class's JavaDoc.

Reply:
No.  That's unfortunate about the ordering.  I was more focused on helping threading.